File: rust/crates/oneproduct-core/src/ontology.rs
//! Shared ontology — the canonical vocabulary every agent uses.
//!
//! Versioning rule: add new fields with `#[serde(default)]` so older agents
//! can still deserialize. Never reuse a field name with a different type.

use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub struct Sku(pub String);

impl From<&str> for Sku {
    fn from(s: &str) -> Self { Self(s.to_string()) }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Product {
    pub sku: Sku,
    pub name: String,
    #[serde(default)]
    pub category: Option<String>,
    #[serde(default)]
    pub unit_cost_try: Decimal,
    #[serde(default)]
    pub list_price_try: Decimal,
    #[serde(default)]
    pub weight_kg: f64,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum OrderStatus {
    Created,
    Paid,
    Shipped,
    Delivered,
    Cancelled,
    Returned,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Order {
    pub order_id: String,
    pub sku: Sku,
    pub quantity: u32,
    pub total_try: Decimal,
    pub status: OrderStatus,
    #[serde(default)]
    pub destination_city: Option<String>,
    #[serde(default)]
    pub created_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum Party { Buyer, Seller, Broker }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Offer {
    pub offer_id: String,
    pub party: Party,
    pub price_try: Decimal,
    #[serde(default = "default_qty")]
    pub quantity: u32,
    #[serde(default)]
    pub moq: Option<u32>,
    #[serde(default)]
    pub lead_time_days: Option<u32>,
    #[serde(default)]
    pub valid_until: Option<DateTime<Utc>>,
}

fn default_qty() -> u32 { 1 }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Invoice {
    pub invoice_id: String,
    pub order_id: String,
    pub total_try: Decimal,
    pub issued_at: DateTime<Utc>,
    #[serde(default)]
    pub paid: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ShipmentQuote {
    pub carrier: String,
    pub cost_try: Decimal,
    pub eta_hours: u32,
    #[serde(default = "default_sla")]
    pub sla_pct: f64,
}

fn default_sla() -> f64 { 0.95 }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PriceSignal {
    pub sku: Sku,
    pub current_price_try: Decimal,
    #[serde(default)]
    pub competitor_avg_try: Option<Decimal>,
    #[serde(default = "default_demand")]
    pub demand_score: f64,
    #[serde(default = "default_stock_days")]
    pub stock_days: i32,
}

fn default_demand() -> f64 { 0.5 }
fn default_stock_days() -> i32 { 30 }

File: backend/apps/api/core/autonomy/ontology.py
"""Ortak ontoloji — multi-agent katmanının paylaşılan kavram sözlüğü.

Her ajan bu Pydantic modellerini referans alır; serileştirme sırasında alan
adları sabit kalır. Bu sayede `negotiation_agent`, `logistics_agent`,
`dynamic_pricing_agent` gibi heterojen ajanlar aynı `Product`, `Order`,
`Offer` kavramı üzerinde konuşur. Yeni alan eklerken default değer ver:
geriye dönük uyumluluk için zorunlu.
"""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, NonNegativeFloat, PositiveInt


class Product(BaseModel):
    sku: str
    name: str
    category: str | None = None
    unit_cost_try: NonNegativeFloat = 0.0
    list_price_try: NonNegativeFloat = 0.0
    weight_kg: NonNegativeFloat = 0.0


class Order(BaseModel):
    order_id: str
    sku: str
    quantity: PositiveInt
    total_try: NonNegativeFloat
    status: Literal["created", "paid", "shipped", "delivered", "cancelled", "returned"] = "created"
    destination_city: str | None = None
    created_at: datetime | None = None


class Offer(BaseModel):
    """Tek bir müzakere teklifi (her iki yöne uygulanabilir)."""

    offer_id: str
    party: Literal["buyer", "seller", "broker"]
    price_try: NonNegativeFloat
    quantity: PositiveInt = 1
    moq: PositiveInt | None = None
    lead_time_days: PositiveInt | None = None
    valid_until: datetime | None = None


class Invoice(BaseModel):
    invoice_id: str
    order_id: str
    total_try: NonNegativeFloat
    issued_at: datetime
    paid: bool = False


class ShipmentQuote(BaseModel):
    carrier: str
    cost_try: NonNegativeFloat
    eta_hours: PositiveInt
    sla_pct: float = Field(ge=0.0, le=1.0, default=0.95)


class PriceSignal(BaseModel):
    sku: str
    current_price_try: NonNegativeFloat
    competitor_avg_try: NonNegativeFloat | None = None
    demand_score: float = Field(ge=0.0, le=1.0, default=0.5)
    stock_days: int = 30
